src/Umbraco.Web.UI.Client/src/utilities.js

```diff
@@ -116,6 +116,254 @@
+
+                return {
+                    mediaEntry,
+                    file
+                };
+            });
+        };
+
+        function validateMaxFileSize (file, val) {
+            return file.size - 0.1 <= Upload.translateScalars(val);
+        }
+    
+        function _upload(queueItem) {
+            const mediaEntry = queueItem.mediaEntry;
+
+            _emit('uploadStarted', { mediaEntry });
+            Upload.upload({
+                    url: uploadURL,
+                    file: queueItem.file
+                })
+                .progress(function(evt) {
+                    var progressPercentage = parseInt(100.0 * evt.loaded / evt.total, 10);
+                    mediaEntry.$uploadProgress = progressPercentage;
+                })
+                .success(function (data) {
+                    _emit('uploadSuccess', { mediaEntry, ...data });
+                    _processQueue();
+                })
+                .error(function(error) {
+                    _emit('uploadError', { mediaEntry });
+                    _rejectMediaEntry(mediaEntry, { type: 'server', message: error.Message });
+                    _processQueue();
+                });
+        }
+
+        function _addItemsToQueue (queueItems) {
+            queue = [...queue, ...queueItems];
+        }
+
+        function _processQueue () {
+            const nextItem = queue.shift();
+
+            // queue is empty
+            if (!nextItem) {
+                _emit('queueCompleted');
+                return;
+            }
+
+            _getMatchedMediaType(nextItem.file).then(mediaType => {
+                nextItem.mediaEntry.mediaTypeAlias = mediaType.alias;
+                nextItem.mediaEntry.$icon = mediaType.icon;
+                _upload(nextItem);
+            }, () => {
+                _rejectMediaEntry(nextItem.mediaEntry, { type: 'pattern', message: translations.disallowedFileType });
+                _processQueue();
+            });
+        }
+
+        function _getMatchedMediaType(file) {
+            return new Promise((resolve, reject) => {
+                const uploadFileExtension = mediaHelper.getFileExtension(file.name);
+                const matchedMediaTypes = mediaTypeHelper.getTypeAcceptingFileExtensions(allowedMediaTypes, [uploadFileExtension]);
+                
+                if (matchedMediaTypes.length === 0) {
+                    reject();
+                    return;
+                };
+                
+                if (matchedMediaTypes.length === 1) {
+                    resolve(matchedMediaTypes[0]);
+                    return;
+                };
+
+                // when we get all media types, the "File" media type will always show up because it accepts all file extensions.
+                // If we don't remove it from the list we will always show the picker.
+                const matchedMediaTypesNoFile = matchedMediaTypes.filter(mediaType => mediaType.alias !== "File");
```

Comment:
I'm having a bit of trouble with this. I have two media types in my solution (File and Article), which coincidentally both accept PDF files. When I'm dragging a PDF over the editor, I'm never asked what it is. I presume it "autopicks" the File type, but in reality it could be the other type. I assume it has something to do with this line where Files are being filtered away.

Conversely, when I drag a PDF into the media library itself, it gives me options to choose between Autopick, File and Article.

Reply:
**A couple of questions:**
* Have you specifically set PDF as allowed extension on File.
* Can you double-check that it didn't autopick the Article type?
* I don't fully understand the purpose of the Auto pick option. When I get a list of available options and I select Auto Pick. Which one will it be?

**This is how I wanted it to work for the media picker:**
The File Media Type should be the fallback type if nothing else fits because it supports everything out of the box. If we find more specific media types that all allows the uploaded file extension, we present the user with a list of those. If we only find one, that is not "File", then that will be the best fit, and we auto select that one.

Reply:
1) Yes
2) It picked the "File" type
3) No I don't know if we should use it here, just wondered why it's in the media library and not here

The dialog simply didn't appear even though I have "PDF" listed on both types. In fact they use the same FileUpload datatype underneath.